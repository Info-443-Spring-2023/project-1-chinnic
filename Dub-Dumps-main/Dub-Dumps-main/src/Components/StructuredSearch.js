import React, { useState } from 'react'; //import React Component
import { Link, Outlet } from 'react-router-dom';
import { Nav, Navbar } from "react-bootstrap";

export function StructuredSearch(props) {
    const [bldgSelected, setBldg] = useState('');
    const [floorSelected, setFloor] = useState('');
    const [locationSelected, setLocation] = useState('');

    const changeBldg = evt => {
        setBldg(evt.target.value);
    }

    const changeFloor = evt => {
        setFloor(evt.target.value);
    }

    const changeLocation = evt => {
        setLocation(evt.target.value);
    }

    const handleClick = () => {
        props.filterCallback(bldgSelected, floorSelected, locationSelected);
    }

    // Returns a unique array of options for a specific filtered value
    function createUniqueOptions(valueType) {
        const allOptions = props.data
            .map(item => item[valueType]);  // map to get an array of only the specific value type

        const uniqueOptions = [...new Set(allOptions)]; // creates Set for unique values, then back to array

        return uniqueOptions.map((valueType) => {
            return <option key={valueType} value={valueType}>{valueType}</option>;
        })
    }

    // Array of buildings
    const buildings = createUniqueOptions("building");

    // Array of floors
    const floors = createUniqueOptions("floor");

    // Array of locations
    const locations = createUniqueOptions("location");

    return (
        <div><Navbar collapseOnSelect expand="lg" bg="dark" variant="dark" name="navbar">
            <Navbar.Toggle aria-controls="responsive-navbar-nav" className="hamburger" />
            <Navbar.Collapse id="responsive-navbar-nav navBar">
                <Nav className="mr-auto">
                    <select id="buildingSelect" className="navDropDown" onChange={changeBldg} data-testid="select building">
                        <option value="">Building</option>
                        {buildings}
                    </select>
                    <select id="floorSelect" className="navDropDown" onChange={changeFloor} data-testid="select floor">
                        <option value="">Floor</option>
                        {floors}
                    </select>
                    <select id="locationSelect" className="navDropDown" onChange={changeLocation} data-testid="select location">
                        <option value="">Location</option>
                        {locations}
                    </select>
                    <div className="col-auto">
                        <Link to="/search" id="submitButton" type="submit" className="btn btn-warning" onClick={handleClick}>Search!</Link>
                    </div>
                </Nav>
            </Navbar.Collapse>
        </Navbar>
            <Outlet />
        </div >
    )
}

export default StructuredSearch;