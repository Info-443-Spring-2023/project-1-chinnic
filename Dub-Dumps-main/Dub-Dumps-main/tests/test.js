import renderer from 'react-test-renderer';
import { BrowserRouter as Router, Routes, Route, Navigate, useNavigate as navigateTo } from 'react-router-dom';
import { render, screen, cleanup } from "@testing-library/react";
import '@testing-library/jest-dom';
import userEvent from "@testing-library/user-event";

import HomePage from '../src/Components/HomePage.js';
import StructuredSearch from '../src/Components/StructuredSearch.js';
import bathroom from '../public/data/bathrooms.json';

// example tests to make sure I was doing it right

describe('Example tests',() => {

    describe('Find A Bathroom! Component',() => {

        it('renders the home page', () => {
            const component = renderer.create(
                <Router>
                    <HomePage></HomePage>
                </Router>
            );
            let tree = component.toJSON();
            expect(tree).toMatchSnapshot();
        });

        test("The text renders", () => {
            render(  
                <Router>
                    <HomePage />
                </Router>
            );
            const text = screen.getByTestId('map-title');
            expect(text).toBeInTheDocument(); 
        })
    
        test("The text content matches", () => {
            render(  
                <Router>
                    <HomePage />
                </Router>
            );
            const text = screen.getByTestId('map-title');
            //expect(text).toHaveTextContent("Find a Bathroom!");
            expect(screen.getByText("Find a Bathroom!")).toBeInTheDocument(); 
        })
    })
})

// 

describe('Structured Searches Component Renders Correctly',() => {
    
    test("navigation search bar renders without errors", () => {
        const callback = jest.fn();
        render(  
            <Router>
                <StructuredSearch data={bathroom} filterCallback={callback} />
            </Router>
        );
        expect(screen.getByRole('navigation')).toBeInTheDocument(); 
    })

    test("snapshot of component renders correctly", () => {
        const callback = jest.fn();
        const component = renderer.create(
            <Router>
                <StructuredSearch data={bathroom} filterCallback={callback} />
            </Router>
        );
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    })
})

describe('Default values are correct',() => {
    
    test("Building default value", () => {
        const callback = jest.fn();
        render(  
            <Router>
                <StructuredSearch data={bathroom} filterCallback={callback} />
            </Router>
        );
        expect(screen.getByRole('option', {name: 'Building'}).selected).toBe(true);
    })

    test("Floor default value", () => {
        const callback = jest.fn();
        render(  
            <Router>
                <StructuredSearch data={bathroom} filterCallback={callback} />
            </Router>
        );
        expect(screen.getByRole('option', {name: 'Floor'}).selected).toBe(true);
    })

    test("Location default value", () => {
        const callback = jest.fn();
        render(  
            <Router>
                <StructuredSearch data={bathroom} filterCallback={callback} />
            </Router>
        );
        expect(screen.getByRole('option', {name: 'Location'}).selected).toBe(true);
    })
})


describe('Allows user to change options',() => {
    
    test("Changing building value", () => {
        const callback = jest.fn();
        render(  
            <Router>
                <StructuredSearch data={bathroom} filterCallback={callback} />
            </Router>
        );
        userEvent.selectOptions(
            screen.getByTestId('select building'),
            screen.getByRole('option', {name: 'RAI'}),
        );
        expect(screen.getByRole('option', {name: 'RAI'}).selected).toBe(true);
    })

    test("Changing floor value", () => {
        const callback = jest.fn();
        render(  
            <Router>
                <StructuredSearch data={bathroom} filterCallback={callback} />
            </Router>
        );
        userEvent.selectOptions(
            screen.getByTestId('select floor'),
            screen.getByRole('option', {name: 'Basement'}),
        );
        expect(screen.getByRole('option', {name: 'Basement'}).selected).toBe(true);
    })

    test("Changing location value", () => {
        const callback = jest.fn();
        render(  
            <Router>
                <StructuredSearch data={bathroom} filterCallback={callback} />
            </Router>
        );
        userEvent.selectOptions(
            screen.getByTestId('select location'),
            screen.getByRole('option', {name: 'South'}),
        );
        expect(screen.getByRole('option', {name: 'South'}).selected).toBe(true);
    })
})

describe('Button Works Correctly',() => {

    test("Button renders correctly", () => {
        const callback = jest.fn();
        render(  
            <Router>
                <StructuredSearch data={bathroom} filterCallback={callback} />
            </Router>
        );
        expect(screen.getByRole('link', {name: 'Search!'})).toBeInTheDocument();
    })

    test("Filter callback function called", () => {
        const callback = jest.fn();
        render(  
            <Router>
                <StructuredSearch data={bathroom} filterCallback={callback} />
            </Router>
        );
        userEvent.click(screen.getByRole('link'));
        expect(callback).toBeCalled();
    })
})