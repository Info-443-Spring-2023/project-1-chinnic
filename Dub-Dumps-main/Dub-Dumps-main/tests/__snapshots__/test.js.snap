// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Example tests Find A Bathroom! Component renders the home page 1`] = `
<div>
  <a
    className="map-link"
    href="/search"
    onClick={[Function]}
  >
    <h2
      className="map-title"
      data-testid="map-title"
    >
      Find a Bathroom!
    </h2>
  </a>
  <div>
    <div
      className="map"
    >
      <iframe
        className="map-image"
        height="550px"
        src="https://snazzymaps.com/embed/398528"
        title="DUB DUMPS bathroom map"
        width="80%"
      />
      <p>
        Use our map to find nearby bathrooms!
      </p>
    </div>
  </div>
</div>
`;

exports[`Structured Searches Component Renders Correctly snapshot of component renders correctly 1`] = `
<div>
  <nav
    className="navbar navbar-expand-lg navbar-dark bg-dark"
    name="navbar"
  >
    <button
      aria-controls="responsive-navbar-nav"
      aria-label="Toggle navigation"
      className="hamburger navbar-toggler collapsed"
      onClick={[Function]}
      type="button"
    >
      <span
        className="navbar-toggler-icon"
      />
    </button>
    <div
      aria-expanded={null}
      className="navbar-collapse collapse"
      id="responsive-navbar-nav navBar"
    >
      <div
        className="mr-auto navbar-nav"
        onKeyDown={[Function]}
      >
        <select
          className="navDropDown"
          data-testid="select building"
          id="buildingSelect"
          onChange={[Function]}
        >
          <option
            value=""
          >
            Building
          </option>
          <option
            value="RAI"
          >
            RAI
          </option>
          <option
            value="ODE"
          >
            ODE
          </option>
          <option
            value="KNE"
          >
            KNE
          </option>
          <option
            value="GWN"
          >
            GWN
          </option>
          <option
            value="CHB"
          >
            CHB
          </option>
          <option
            value="ALB"
          >
            ALB
          </option>
          <option
            value="LEW"
          >
            LEW
          </option>
          <option
            value="PCAR"
          >
            PCAR
          </option>
          <option
            value="IMA"
          >
            IMA
          </option>
          <option
            value="MGH"
          >
            MGH
          </option>
          <option
            value="WIL"
          >
            WIL
          </option>
          <option
            value="GLD"
          >
            GLD
          </option>
          <option
            value="LAW"
          >
            LAW
          </option>
          <option
            value="HUB"
          >
            HUB
          </option>
          <option
            value="PAB"
          >
            PAB
          </option>
          <option
            value="PDL"
          >
            PDL
          </option>
          <option
            value="BRK"
          >
            BRK
          </option>
          <option
            value="ART"
          >
            ART
          </option>
          <option
            value="SUZ"
          >
            SUZ
          </option>
          <option
            value="JHN"
          >
            JHN
          </option>
          <option
            value="SAV"
          >
            SAV
          </option>
          <option
            value="LOW"
          >
            LOW
          </option>
          <option
            value="LSB"
          >
            LSB
          </option>
          <option
            value="WFS"
          >
            WFS
          </option>
          <option
            value="BLD"
          >
            BLD
          </option>
          <option
            value="MOR"
          >
            MOR
          </option>
        </select>
        <select
          className="navDropDown"
          data-testid="select floor"
          id="floorSelect"
          onChange={[Function]}
        >
          <option
            value=""
          >
            Floor
          </option>
          <option
            value="First Floor"
          >
            First Floor
          </option>
          <option
            value="Second Floor"
          >
            Second Floor
          </option>
          <option
            value="Third Floor"
          >
            Third Floor
          </option>
          <option
            value="Basement"
          >
            Basement
          </option>
          <option
            value="Fourth Floor"
          >
            Fourth Floor
          </option>
        </select>
        <select
          className="navDropDown"
          data-testid="select location"
          id="locationSelect"
          onChange={[Function]}
        >
          <option
            value=""
          >
            Location
          </option>
          <option
            value="North"
          >
            North
          </option>
          <option
            value="West"
          >
            West
          </option>
          <option
            value="South"
          >
            South
          </option>
          <option
            value="East"
          >
            East
          </option>
        </select>
        <div
          className="col-auto"
        >
          <a
            className="btn btn-warning"
            href="/search"
            id="submitButton"
            onClick={[Function]}
            type="submit"
          >
            Search!
          </a>
        </div>
      </div>
    </div>
  </nav>
</div>
`;
